### ISSUE TITLE
Keras subclassing and explicit dtype of Input

### ISSUE THREAD

**Issue by perara** [tensorflow/tensorflow#27705]
**System information**

Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
**OS Platform and Distribution (e.g., Linux Ubuntu 16.04):** Arch Linux
**Tensorflow Version:** 2.0.0-alpha0
**Description**
When using Keras subclassing there is no apparent way of defining the dtype of the Input node of the network. In some cases, it would be neccecary to use tf.float16 instead of 32 but as of now i cannot find any way to adjust this. Also trying to set the dtype using self.dtype = tf.float16 is not permitted.

**Comment by achandraa** [comment]
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by perara** [comment]
**Method 1:**
```
import tensorflow as tf


class SomeClass(tf.keras.Model):

    def __init__(self, dtype):
        super(SomeClass, self).__init__(dtype=dtype)  # Set Input of dtype either like this

        self.h_1 = tf.keras.layers.Dense(10, dtype=dtype)

    def call(self, inputs):
        return self.h_1(inputs)


if __name__ == "__main__":

    model = SomeClass(dtype=tf.float16)
    model([1, 2, 0])
```
Exception:
```
TypeError: _init_subclassed_network() got an unexpected keyword argument 'dtype'
```

**Method 2**
```
import tensorflow as tf


class SomeClass(tf.keras.Model):

    def __init__(self, dtype):
        self.dtype = dtype  # Or this?

        self.h_1 = tf.keras.layers.Dense(10, dtype=dtype)

    def call(self, inputs):
        return self.h_1(inputs)


if __name__ == "__main__":

    model = SomeClass(dtype=tf.float16)
    model([1, 2, 0])
```
Exception:
```
AttributeError: can't set attribute
```

**Use case**
The use case is when you want specifically to use custom dtypes for the whole net, typically when running on TensorCores etc.

**Freetext**
I dont know if this would be the correct way of defining this, but it should be possible (if its not already) to set the dtype of the input excplictly ( i assume there is a tf.cast in there somewhere anyways...)

**Comment by k-w-w** [comment]
Hi! A workaround you can do now is to directly set the private property `_dtype`. I'll make a change that adds `dtype` as one of the allowed keyword arguments super().__init__.

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved, but please let me know if I'm mistaken.Thanks!

**Comment by jvishnuvardhan** [comment]
Reopening as I find some issues are there still with `tf-nightly-gpu-2.0-preview==2.0.0.dev20190723` . Here is the [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/4ee37c87967ef4abfa7c7cb3bed63c0c/tf_27705.ipynb). Thanks!

**Comment by ymodak** [comment]
For method 2:
dtype is a property therefore we see error
The error message is fixed with latest tf-nightly 2.0 build version '2.0.0-dev20190809'
```python
AttributeError: Can't set the attribute "dtype", likely because it conflicts with an existing read-only @property of the object. Please choose a different name..
```

**Comment by k-w-w** [comment]
@jvishnuvardhan Use method 1. 

```
ValueError: Layer dense expects 1 inputs, but it received 3 input tensors. Inputs received: [<tf.Tensor: id=1, shape=(), dtype=int32, numpy=1>, <tf.Tensor: id=2, shape=(), dtype=int32, numpy=2>, <tf.Tensor: id=3, shape=(), dtype=int32, numpy=0>]
```

The above error is appearing because [1, 2, 0] is being converted into separate tensors. Using `tf.constant` to convert the entire array into a tensor (try `model(tf.constant([[1, 2, 0]]))`).
